### DISCUSSION TITLE
Server Actions & CSRF Tokens

### DISCUSSION THREAD

**Discussion by coltonehrman** [vercel/next.js#58543] [Help]
### Summary

There is an error we get when trying to parse a server action multipart request before the Next.js internal multipart **busboy** parser.

<img width="987" alt="Screenshot 2023-11-16 at 10 17 09 AM" src="https://github.com/vercel/next.js/assets/12456288/20210aad-4a35-4bdd-bccf-9d9a28c84b97">

This leads me to a couple of questions listed below.

1. What is the recommended pattern for enabling CSRF protection on server action forms?
2. Can we change the form type of a server action form?
3. Can we pre-parse the req.body of a server action form?
4. How should I handle manage multipart parsing conflicts in my application if I have my own multipart parser enabled on forms?
5. Is it possible to include additional headers (ie: CSRF) as part of a server action form?

### Why can’t we use our own multipart parser?

Because of Node.js streams (which is where the request form data comes from), it is not possible (AFAIK) to read/consume from a stream more than once. As soon as you consume data from a stream and it has ended, trying to read from the stream again will result in an error.

There are some “possible” ways to push the data back into a stream and enable reading again later, or even rewind the stream back, but these solutions are not reliable and prone to errors. Would not recommend this.

**Comment by shaunwarman** [comment]
cc @timneutkens @leerob

**Comment by Tommy1223454** [comment]
A practical pattern here is to keep CSRF/auth concerns around the form submission, but move raw file ingestion into a dedicated upload endpoint so you don’t have competing multipart parsers on the same request stream.

Then the upload route can inspect the file before storage / downstream processing, and the server action can deal with metadata or a promoted file reference instead of the raw multipart body.

I built Pompelmi for this exact Node.js layer:
https://github.com/pompelmi/pompelmi

It’s basically an inspect-before-storage step for untrusted uploads (type sniffing, MIME/extension mismatch checks, archive abuse signals, etc.). Not saying it solves the parser conflict itself, but it fits well once you split the pipeline into “ingest” vs “business action”.
